Feature modelpart interface constraints

adityaghantasala opened the pull request:
Hi @RiccardoRossi and @pooyan-dadvand 
This pull request its meant to aid and accelerate the process of adding master slave constraints to modelpart. 

It retains the previous implementation of storing the constraint (using key and ID) for time being. 

Major changes in the Modelpart class and the residual_based_block_builder_and_solver.h

I also added the MasterSlaveConstraint base class and exposed it to the python using the corresponding kratos/python/add_constraint_to_python.h/cpp

The technical implementation of the constraint itself remains the same as previous but will be called/implemented via this interface.

adityaghantasala committed:
Adding class with data structure to store master slave relation s = T*m + c

adityaghantasala committed:
Adding the base constraint class which contains a constraint equation.

adityaghantasala committed:
TODO to master slave relation class

adityaghantasala committed:
Initial changes to the model_part , adding type defs for constraints

adityaghantasala committed:
renaming constrant class to MasterSlaveConstraint

adityaghantasala committed:
initial changes to add vector of constraints to modelpart

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Adding add and create function for the master-slave constraints

adityaghantasala committed:
Adding remove function declarations for master-slave constraints

adityaghantasala committed:
Adding get methods for master-slave constraint

adityaghantasala committed:
adding missing const for functions, and adding overload of << and >> operators

adityaghantasala committed:
Adding definition of the functions for master-slave constraints. The functions are empty for now

adityaghantasala committed:
adding scope operators for compilation

adityaghantasala committed:
Adding missing empty constructor and create funtion

adityaghantasala committed:
Adding function definitions (implementing). Changing the constraint container to a set from a map

adityaghantasala committed:
Formatting issues

adityaghantasala committed:
Adding initialize and finalize solution step functions to constrint base class

adityaghantasala committed:
Adding constraint base class and the corresponding addition to python

adityaghantasala committed:
Adding modelpart as argument for modification functions in builder and solver

adityaghantasala committed:
Merge branch 'core/uniform-def-of-resize-and-initialize' into feature-modelpart-interface-constraints

adityaghantasala committed:
Adding MasterSlaveConstraintsBegin function

adityaghantasala committed:
changes to construct matrix structure to use modelpart

adityaghantasala committed:
Using ApplyConstraints in the builder and solver at appropriate places

adityaghantasala committed:
Adding MasterSlaveConstraint interface to ModelPart , related addition to python, small bug fixes and style improvements

adityaghantasala committed:
Adding the MasterSlaveConstraint base class to python

adityaghantasala committed:
clean up in the add constraint to python cpp file

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Adding create and add master-slave constraints functions to python level.

adityaghantasala committed:
removing trailing spaces from the test_model_part.py

adityaghantasala committed:
Adding AddSlave method and some name changes and formatting

adityaghantasala committed:
Adding and implementing AddSlave function and updating to changed interface of master slave relation class

adityaghantasala committed:
Simplifying the definition of the function CreateNewMasterSlaveFunction

adityaghantasala committed:
Adding MasterSlaveConstraint Object along side with different elements and conditions

adityaghantasala committed:
removing second constructor from python for constraint class

adityaghantasala committed:
Adding complete and over loaded implementation of the AddSlave and AddMaster to Constraint class

adityaghantasala committed:
Bug fixes and complete definition in AddMaster and AddSlave in Constraint.h

adityaghantasala committed:
Defining KRATOS_REGISTER_CONSTARINT macro in define.h

adityaghantasala committed:
Adding, registering and defining the constraint base class in varibles.h

RiccardoRossi commented on kratos/includes/model_part.h:
```diff
@@ -208,6 +209,33 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+    /**
+     * Type of the weight for a master
+     * */
+    typedef double MasterWeightType;
```
do we need this?

RiccardoRossi commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +596,94 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+        return mMasterSlaveConstraints;
+    }
+
+    void SetMasterSlaveConstraints(MasterSlaveConstraintContainerType::Pointer pOtherMasterSlaveConstraints)
```
i would propose not to define this function. actually i would like to remove the similar one also for the Nodes, Elements and Conditions...(this last one is out of scope here, but just to motivate my comment)

RiccardoRossi commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +596,94 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    /** Inserts an master-slave constraint in the current mesh.
+     */
+    MasterSlaveConstraintType::Pointer CreateNewMasterSlaveConstraint(std::string ConstraintName, IndexType SlaveNodeId);
```
why just one SlaveNodeId?

RiccardoRossi commented on kratos/python/add_constraint_to_python.cpp:
```diff
@@ -0,0 +1,47 @@
+{
+
+    class_<MasterSlaveConstraint>(m, "MasterSlaveConstraint" )
+	.def(init<>())
```
i think we should expose a richer constructor

RiccardoRossi replied:
i mean somethign like

     MasterSlaveConstraint( 
          master_dofs. 
          slave_dofs,
          T,
          b
       )

RiccardoRossi commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,196 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+        else //master-slave constraint did exist already
+        {
+            // In this case, we should update the equation with possible additional masters in the new incoming constraint. 
+            if(&(*existing_condition_it) != (pNewMasterSlaveConstraint.get()))//check if the pointee coincides
```
nope, in this case you should throw an error...

adityaghantasala committed:
Changing author name

RiccardoRossi commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,196 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+/** Inserts an master-slave constraint in the current mesh.
+ */
+ModelPart::MasterSlaveConstraintType::Pointer ModelPart::CreateNewMasterSlaveConstraint(std::string ConstraintName, 
+                                                                                        ModelPart::IndexType SlaveNodeId)
```
we need to think about this

RiccardoRossi commented on kratos/utilities/constraint.h:
```diff
@@ -0,0 +1,279 @@
+     * @param pProperties the properties assigned to the new element
+     * @return a Pointer to the new element
+     */
+    virtual Pointer Create(IndexType NewId) const
```
I would not provide this kind of construction. In my opinion we should always give it as fully constructed

RiccardoRossi reviewed (commented):
a few comments

adityaghantasala committed:
Adding new and rich constructors for constraint base class. Also corresponding create functions.
Removing the object of the constraint equation class from constraint base class and necessary changes.
Removing AddSlave and AddMaster functions from Constraint base class.

adityaghantasala committed:
Adding MasterSlaveConstraint object to kratos applicaiton class and necessary changes to this file.

adityaghantasala committed:
Overloading CreateNewMasterSlaveConstraint function to use new constructors of the MasterSlaveConstraint Class

adityaghantasala committed:
Small changes to make it similar to other files structure

adityaghantasala committed:
Calling AddConstraintsToPython along with other

adityaghantasala committed:
From ModelPart class exposing different way of adding Constraints to python.

adityaghantasala committed:
removing MasterWeightType typedef -- unused

adityaghantasala committed:
Deriving constraint equation from IndexedObject and necessary changes in the class and its constructors

adityaghantasala committed:
Adding object of global master slave constraints to builder and sovler
Adding FormulateGlobalMasterSlaveRelations function (minimal) for computing the global relation

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -521,6 +537,39 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
+        EquationIdVectorType  slave_equation_ids = EquationIdVectorType(0);
+        EquationIdVectorType  master_equation_ids = EquationIdVectorType(0);
+
+        for (int i_constraints = 0; i_constraints < number_of_constraints; i_constraints++)
```
good, tbd in parallel in the future though ;-)

msandre commented on kratos/includes/model_part.h:
```diff
@@ -103,6 +108,11 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
 
     typedef Dof<double> DofType;
+    typedef std::vector< DofType::Pointer > DofsVectorType;
+    typedef Kratos::Variable<double> VariableType;
```
`DoubleVariableType` would be more descriptive.

msandre commented on kratos/includes/model_part.h:
```diff
@@ -208,6 +218,27 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+    typedef PointerVectorSet<MasterSlaveConstraintType, IndexedObject> MasterSlaveConstraintContainerType;
+    //typedef PointerVectorSet<DofType, MasterSlaveConstraintType> MasterSlaveConstraintContainerType;
+
+    /** Iterator over the constraints. This iterator is an indirect
```
I get what this comment is trying to say, but 

> turn back a reference to Table

is confusing.

msandre commented on kratos/includes/model_part.h:
```diff
@@ -208,6 +218,27 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    /** Const iterator over the constraints. This iterator is an indirect
+    iterator over MasterSlaveConstraint::Pointer which turn back a reference to
+    Table by * operator and not a pointer for more convenient
```
Here I almost think the code would be simpler without the comments. This is repeating the last one which was already quite long for a typedef.

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    MasterSlaveConstraintContainerType::Pointer pGetMasterSlaveConstraints()
+    {
+        return Kratos::shared_ptr<MasterSlaveConstraintContainerType>(&mMasterSlaveConstraints);
```
I don't think you should create a shared pointer like this.

adityaghantasala replied:
This functions is necessary basically for making the MasterSlaveConstraints a property of ModelPart when adding it to python. I will see if I can replace it. 
Is there any other way to so this pointer creation.

msandre replied:
The problem is shared_ptr is for heap allocated objects. And even if mMasterSlaveConstraints were heap allocated, one should not create a shared_ptr from raw ptr like this as the count will be wrong. As soon as that ptr goes out of scope something bad will happen.

RiccardoRossi replied:
+1 for Mike's commnet. The current code is simply wrong...

RiccardoRossi replied:
about the python interface `mMasterSlaveConstraints` exists and has a owner who manages its lifetime no?

if that is so, why don't you give back to python a reference instead of a Pointer? (return_internal_reference)

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    void SetMasterSlaveConstraints(MasterSlaveConstraintContainerType::Pointer pOtherMasterSlaveConstraints)
+    {
+        mMasterSlaveConstraints = *pOtherMasterSlaveConstraints;
```
Also here I think you should pass by reference in the function parameter, not by shared pointer since you are not passing ownership.

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    /** Creates a new master-slave constraint in the current modelpart.
+     */
+    MasterSlaveConstraintType::Pointer CreateNewMasterSlaveConstraint(std::string ConstraintName, IndexType Id, DofsVectorType& rMasterDofsVector,
```
Why do you define a `ConstraintName` when you already have a `Id`?

msandre replied:
If you decide it's necessary, it would be better to pass string by const&.

RiccardoRossi replied:
+1 for the const. The idea of passing the string is that you can have different constraint types as you have different element types

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+    MasterSlaveConstraintType::Pointer CreateNewMasterSlaveConstraint(std::string ConstraintName, IndexType Id, DofsVectorType& rMasterDofsVector,
+                                                                                    DofsVectorType& rSlaveDofsVector,
+                                                                                    MatrixType RelationMatrix,
+                                                                                    VectorType ConstantVector);
```
Can we pass the matrix and vector by const&?

RiccardoRossi replied:
+1 for the const

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    /** Remove given master-slave constraint from mesh with ThisIndex in this modelpart and all its subs.
+    */
+    void RemoveMasterSlaveConstraint(MasterSlaveConstraintType::Pointer pThisMasterSlaveConstraint);
```
I would try to avoid this function and just leave the one with a reference to the constraint.

RiccardoRossi replied:
+1

msandre commented on kratos/includes/model_part.h:
```diff
@@ -568,6 +599,111 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    /** Remove the master-slave constraint with given Id from mesh with ThisIndex in parents, itself and children.
+    */
+    void RemoveMasterSlaveConstraintFromAllLevels(IndexType MasterSlaveConstraintId);
```
Can we avoid this function by doing `GetRootModelPart().RemoveMasterSlaveConstraint()`?

RiccardoRossi replied:
Makes sense to me, however Elements and Conditions have this same function, so for "symmetry" we should provide it (or remove it also for elements and conditions -- which i would definitely discuss in a different PR)

pooyan-dadvand replied:
Yes, but it is not what all the developers can deduce by themselves. FromAllLevels gives them better description than the  replacing code.

msandre commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,291 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+*/
+void ModelPart::RemoveMasterSlaveConstraint(ModelPart::MasterSlaveConstraintType& ThisMasterSlaveConstraint)
+{
+    mMasterSlaveConstraints.erase(ThisMasterSlaveConstraint.Id());
```
You can avoid code redundancy by doing `RemoveMasterSlaveConstraint(ThisMasterSlaveConstraint.Id());`

msandre commented on kratos/utilities/constraint.h:
```diff
@@ -0,0 +1,297 @@
+                                        DofsVectorType& rSlaveDofsVector,
+                                        MatrixType RelationMatrix,
+                                        VectorType ConstantVector):IndexedObject(Id), Flags()
+    {
```
Why define the constructor with unused parameters?

adityaghantasala replied:
@msandre for now these are only prototypes of constructors as suggested by @RiccardoRossi . I am currently implementing those.

msandre replied:
ok

msandre commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:
```diff
@@ -0,0 +1,639 @@
+///@{
+
+/**
+ * @class ResidualBasedEliminationBuilderAndSolver
```
Wrong class name.

msandre commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,255 @@
+namespace Kratos
+{
+
+// MasterSlaveRelation class start
```
Try to use the doxygen `@class` instead.

msandre reviewed (commented):
I think I get it now why you have the string and the id. Just ignore that comment.

adityaghantasala committed:
Adding doxygen style comments for MasterSlaveRelation class. Simplyfying the class to contain only a map

adityaghantasala committed:
formatting

adityaghantasala committed:
addressing reviews from github. mostly minor changes and formatting

adityaghantasala committed:
Adding definitions to the constructor of constraint base class

adityaghantasala committed:
Adding calculate local system methods to constraint equation class

RiccardoRossi commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,291 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+    }
+    else
+    {
+        auto existing_condition_it = MasterSlaveConstraints().find(pNewMasterSlaveConstraint->Id());
```
i would call it existing_constraint_it

adityaghantasala committed:
Adding initial implementation to make the global constraint vector

adityaghantasala committed:
Adding const for methods in modelpart and changing the function definitions

adityaghantasala commented:
@RiccardoRossi  I also added the data structure necessary to make the global constraints inside the builder and solver. For now this is class around a map of maps to return the matrices and vectors.

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Modifications for compilation errors and warnings

RiccardoRossi commented on kratos/containers/constraint_equation.h:
```diff
@@ -126,69 +69,76 @@ class MasterSlaveRelation : public IndexedObject
-    std::size_t SlaveEquationId() const { return mSlaveEquationId; }
-    void SetSlaveEquationId(std::size_t Id) { mSlaveEquationId = Id; }
+    IndexType SlaveEquationId() const { return mSlaveEquationId; }
 
```
if it is only for the serializer make it private. (the serializer as a friend class will be able to see it)

adityaghantasala replied:
Sure .. !

RiccardoRossi commented on kratos/containers/constraint_equation.h:
```diff
@@ -126,69 +69,76 @@ class MasterSlaveRelation : public IndexedObject
+     */
+    virtual void EquationIdVector(EquationIdVectorType& rSlaveEquationIds,
+                                  EquationIdVectorType& rMasterEquationIds,
+                                  ProcessInfo& rCurrentProcessInfo)
```
to be clean, also create a const version which does not resize

RiccardoRossi commented on kratos/containers/constraint_equation.h:
```diff
@@ -126,69 +69,76 @@ class MasterSlaveRelation : public IndexedObject
+    virtual void CalculateLocalSystem(MatrixType& rTransformationMatrix,
+                                      VectorType& rConstantVector,
+                                      ProcessInfo& rCurrentProcessInfo)
     {
```
const version (without resizing...you throw an error in debug if the size is not correct)

adityaghantasala replied:
I am confused here ... is there a use of this const version ? I was of the understanding that CalculateLocalSystem need not always get matrices and vectors of correct sizes. And these are modified inside this function to suite element/condition/constraint requirement !

RiccardoRossi commented on kratos/containers/constraint_equation.h:
```diff
@@ -203,47 +153,19 @@ class MasterSlaveRelation : public IndexedObject
-            rSerializer.save("master_key", master_data->MasterDofKey()); // saving the id of the master
-            rSerializer.save("weight", master_data->MasterWeight());     // saving the id of the master
-        }
+
```
save and load should be private

RiccardoRossi reviewed (commented):
some more comments

adityaghantasala commented:
@RiccardoRossi  the part about passing reference : I already changed it locally , I will push it together with upcoming changes.

adityaghantasala committed:
moding load, save and add master to private

adityaghantasala committed:
Namespace declartion change

adityaghantasala committed:
Changing pGetMPC function

adityaghantasala committed:
Removing pGetMasterSlaveModelPart from ModelPart

adityaghantasala committed:
Implementation of getEquationIds and getConstraintMatrices functions

adityaghantasala committed:
Implementation of ApplyConstraints for adding node Ids

adityaghantasala committed:
add full implementation of the ApplyConstraints function to modify the equation ids

adityaghantasala committed:
Adding initial implementation for modifying the lhs and rhs of the element with master slave constraints

pooyan-dadvand commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,180 @@
+ *   This equation is imposed on the linear system of equations either element wise or on the global system.
+ * @author Aditya Ghantasala
+ */
+class MasterSlaveRelation : public IndexedObject
```
It would be better to have the same name for class and file name.

pooyan-dadvand commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,180 @@
+    }
+
+
+    MasterSlaveRelation(IndexType const &rSlaveEquationId) : IndexedObject(rSlaveEquationId), mSlaveEquationId(rSlaveEquationId)
```
If the id of the constrain is the one of the slave, why we should save both of them?

pooyan-dadvand commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,180 @@
+
+    virtual void save(Serializer &rSerializer) const override
+    {
+
```
Would you please also implement the save and load?

pooyan-dadvand commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,180 @@
+    ///@}
+
+
+    std::unordered_map<IndexType, double> mMasterDataSet;
```
Is it really needed to be a unordered_map? an std::vector wouldn't be better? consider that brute force search in a small vector is faster than hash. And uses less memory.

pooyan-dadvand commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,180 @@
+
+    IndexType mSlaveEquationId;
+    double mConstant;
+    double mConstantUpdate;
```
What is the ConstantUpdate?

adityaghantasala committed:
corrections for compilation and moving the AddMaster to public interface

adityaghantasala committed:
Implementing all the necessary function for completing the implementation.
And corrections for compilation errors

adityaghantasala committed:
Adding add_constraints_to_python.cpp to sources list

adityaghantasala committed:
Adding set and get data functions for setting bool type of variables (Flags)

adityaghantasala committed:
Adding set and get to for flags

adityaghantasala committed:
exposing setData and GetData function for accessing variables/flags

pooyan-dadvand commented on kratos/utilities/constraint.h:
```diff
@@ -0,0 +1,334 @@
+//    |  /           |
```
shouldn't be master_slave_constraint.h?

pooyan-dadvand reviewed (commented):
My major design concern here is the unclear/undefined relation between MasterSlaveRelation and MasterSlaveConstraint. It seems that MasterSlaveRelation is not even used.

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

pooyan-dadvand commented:
> I think I get it now why you have the string and the id. Just ignore that comment.

So would you please explain it to me? I have not get it... 🤔

adityaghantasala committed:
Removing the duplicate mSlaveEquationId as a member variable

adityaghantasala committed:
renaming the file constraint.h to master_slave_constraint.h and necessary changes

adityaghantasala committed:
renaming the constraint_equation.h to master_slave_relation.h and necessary changes

adityaghantasala committed:
changing the includes name

adityaghantasala committed:
removing mConstantUpdate from master_slave_relation.h

adityaghantasala committed:
removing functions using mConstantUpdate

adityaghantasala committed:
removing more functions using constant update

RiccardoRossi commented on kratos/python/add_model_part_to_python.cpp:
```diff
@@ -464,9 +464,9 @@ ModelPart::MasterSlaveConstraintType::Pointer CreateNewMasterSlaveConstraint2(Mo
                                                                                        ModelPart::IndexType Id,
                                                                                        ModelPart::NodeType& rMasterNode,
-                                                                                       ModelPart::VariableType& rMasterVariable,
+                                                                                       ModelPart::DoubleVariableType& rMasterVariable,
```
what is the variable is a component?

DISPLACEMENT_X to make an example

RiccardoRossi replied:
also i am not really sure of what this Create function is intended to do...

RiccardoRossi commented on kratos/containers/constraint_equation.h:
```diff
@@ -0,0 +1,198 @@
+                                  ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rSlaveEquationIds.size() != 0)
+            rSlaveEquationIds.resize(1);
```
this doesn't need to be 1...

RiccardoRossi replied:
i mean you can connect M slaves to N masters by the corresponding matrix & vector

adityaghantasala committed:
bug fix

RiccardoRossi commented on kratos/utilities/master_slave_constraint.h:
```diff
@@ -0,0 +1,368 @@
+    */
+    MasterSlaveConstraint(IndexType Id, DofPointerVectorType& rMasterDofsVector,
+                                        DofPointerVectorType& rSlaveDofsVector,
+                                        MatrixType& rRelationMatrix,
```
const for the Matrix and the Vector

RiccardoRossi commented on kratos/utilities/master_slave_constraint.h:
```diff
@@ -0,0 +1,368 @@
+
+    virtual void save(Serializer &rSerializer) const override
+    {
+
```
TODO: serialize the private data

RiccardoRossi replied:
(and also the mData)

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,197 @@
+//    ' /   __| _` | __|  _ \   __|
```
this is an implementation detail, so i think it belongs in the same directory as the builderandsolver which uses it. 
(please tell so in the comment describing the object)

adityaghantasala replied:
I put detailed description and comments about this class in the comments.

RiccardoRossi replied:
i think that this one as "implentation detail" of the builder and solver does not belong in this PR

also i would use a name that suggests immediately that this is internal (maybe putting it in a namespace "Internal" or "detail" or something of the sort

RiccardoRossi commented on kratos/includes/model_part.h:
```diff
@@ -574,6 +605,106 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+        return mMasterSlaveConstraints.size();
+    }
+
+    MasterSlaveConstraintContainerType& MasterSlaveConstraints()
```
make this const

adityaghantasala replied:
What if someone wants to loop over and change the constraint. !! The API does not exist now but in future ?

RiccardoRossi replied:
you are right, however there should be both the const and nonconst version as

      MasterSlaveConstraintContainerType& MasterSlaveConstraints()
      const MasterSlaveConstraintContainerType& MasterSlaveConstraints() const

the second one is important if you just have a const ModelPart and you want to be able to call the function

RiccardoRossi commented on kratos/python/add_model_part_to_python.cpp:
```diff
@@ -447,6 +447,93 @@ void ModelPartRemoveConditionsFromAllLevels(ModelPart& rModelPart, Flags identif
+    return rModelPart.MasterSlaveConstraints();
+}
+
+void ModelPartSetMasterSlaveConstraints1(ModelPart& rModelPart, ModelPart::MasterSlaveConstraintContainerType::Pointer pOtherMasterSlaveConstraints)
```
i wonder if it would not be possible to make a perfect forwarding of the arguments to reduce the number of theses functions. Indeed construction here is more complicated than for the case of Elements and Conditions since there are multiple ways to construct the MPC (and there can be many more not contemplated here)

adityaghantasala replied:
@RiccardoRossi  thank you for that suggestion, in this case I  tried it like this 

``` 
template <typename ... Args>
ModelPart::MasterSlaveConstraintType::Pointer CreateNewMasterSlaveConstraint(ModelPart& rModelPart, Args&& ... args)
{
    return rModelPart.CreateNewMasterSlaveConstraint(std::forward<Args>(args) ...);
}
```

But unfortunately the pybind could not deduce Args and args !!

RiccardoRossi replied:
I think that in exporting to pybind you'll have to explicitly add helper functins, however this way of using could still be useful on the c++ side

RiccardoRossi commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,291 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+{
+    if (IsSubModelPart())
+    {
+        mpParentModelPart->AddMasterSlaveConstraint(pNewMasterSlaveConstraint);
```
it should also add it here, not just to the parent. (If you add to a submodelpart it belongs to the subpart AND to the parent)

adityaghantasala replied:
done

RiccardoRossi commented on kratos/sources/variables.cpp:
```diff
@@ -973,6 +977,8 @@ void KratosApplication::RegisterVariables() {
     Serializer::Register("Node3D", NodeType());
     Serializer::Register("DofDouble", Dof<double>());
 
+    Serializer::Register("MasterSlaveConstraint", MasterSlaveConstraint());
```
I think we shall also register it in the KratosComponents. Isn't this handled by the macro you added in the define?

adityaghantasala replied:
Yes I think so, it is done by the macro.

RiccardoRossi reviewed (commented):
getting there :-)

some more comments here and there

RiccardoRossi commented on kratos/utilities/master_slave_constraint.h:
```diff
@@ -0,0 +1,368 @@
+//    |  /           |
```
forgot to tell that this should be a first class citizen, so i guess it goes in includes with Element.h and Condition.h

adityaghantasala committed:
moving master_slave_constraint.h to includes

adityaghantasala committed:
removing the constraint builder and solver from this branch and moving it to another branch

adityaghantasala commented:
@RiccardoRossi  I moved the builder and solver from this branch. 

@pooyan-dadvand  thank you for your comments. I addressed all of them. Name changes are done. 

Regarding 

MasterSlaveConstraint : 

- This is a base class for any further implementations of the master-slave constraints.
- Objects of these class are citizens of the modelpart. They are stored in a PointerVectorSet. 
- The interface of this added to python. Model part also has functions to add an object of MasterSlaveConstraint to the PointerVectorSet. 
- This object can be constructed with a vector of slaves, masters and their relation matrix and vector.
- This object can also be constructed with just one slave and master, corresponding weight and constant. 
- There can be two of these objects with same slave and different masters (corresponding), their weights. 

MasterSlaveRelation:

-  This stores a relation s = T*m+c which condenses all the MasterSlaveConstraint objects. 
-  Objects of this class are created and are available only in BuilderAndSolver. 
-  For now there is possibility of having s = sigma(w_i*m_i) + c , that is one object per slave. 
-  Regarding your question about the unordered_map ... I put a unordered map because, I wanted a unique set of masters in the equation.
- Once objects of this class are created in builder and solver there are only used but not the MasterSlaveConstraint objects.

pooyan-dadvand commented:
Thanks for the clarification!

So why we need the MasterSlaveConstraint to represent several equations at once? we may have a only one equation per constrain and reduce a lot the infrastructure for storing them. 

about the unordered_map I would suggest to use a temporary one (or just test by brute force) and store everything in a vector which would be more efficient and faster.

adityaghantasala committed:
removing addition of constraints builder and solver into python

adityaghantasala commented:
@pooyan-dadvand  I forgot to explain why I need the string. It is to add a particular type of constraint. Just like adding "Triangle2D3N" by specifying its name one can also register a constraint's name and add this type of constraint using its name.

adityaghantasala commented:
@pooyan-dadvand  regarding unordered map, I will implement such a mechanism as you suggested.

adityaghantasala committed:
Adding the constraint to the current modelpart also

RiccardoRossi commented:
@pooyan-dadvand i'll chime in on this. You may have a multipoint constraint having multiple target dofs related to multiple slave dofs. 

The simple case i can think of is that of a slip boundary conditions. (or of a node contact).
This will be preprocessed internally to an auxiliary data structure before being actually built into the builder and solver, however this is an implementation detail and is transparent to the end user

adityaghantasala committed:
correcting the functions that are added to the python interface regarding the master-slave constraints

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+    }
+
+
+    MasterSlaveRelation(IndexType const &rSlaveEquationId) : IndexedObject(rSlaveEquationId)
```
do we need this "empty" constructor?

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    // This is only for serializer. Not to be used outside
+    void AddMaster(std::size_t MasterEquationId, double Weight)
```
private

adityaghantasala replied:
I think there is no need of serializing this class, as the objects of this class are to be only used in b&s and can be reconstructed  any time from the objects of MasterSlaveConstraint (the one which is public). 

But any way, during implementing the builder and solver, I had to have this function as there should be a possibility to add a master to an existing relation. 
I will remove that comment which is wrong.

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+
+    void Clear()
+    {
+        mMasterDataSet.clear();
```
why not also the data?

adityaghantasala replied:
There is no data in this class except this mMasterDataSet.

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+                                      VectorType& rConstantVector,
+                                      ProcessInfo& rCurrentProcessInfo)
+    {
+      if (rTransformationMatrix.size1() != 0)
```
! this should include the case of multiple master dofs...

RiccardoRossi replied:
i mean multiple slaves and multiple masters

adityaghantasala replied:
Yes ... I will add this constructor also.

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+
+} // namespace Kratos
+
+#endif // CONSTRAINT_SLAVE_H_INCLUDED
```
you should add one space at the end

RiccardoRossi commented on kratos/tests/test_model_part.py:
```diff
@@ -5,8 +5,8 @@
-    
+class TestModelPart(KratosUnittest.TestCase):
+
     def setUp(self):
```
we need tests of mpcs...as i see there is nothing added here

adityaghantasala replied:
Added in the latest commit.

RiccardoRossi commented on kratos/containers/master_slave_relation.h:
```diff
@@ -0,0 +1,201 @@
+     * @param rResult the elemental equation ID vector
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    virtual void EquationIdVector(EquationIdVectorType& rSlaveEquationIds,
```
if you know that you can have just one slave, just return one id and not a vector

adityaghantasala replied:
In that case, the builder and solver which is using this function will have to implement two different versions : one where only one equation Id of slave is returned, or a vector is returned ? 
Wouldn't it become a bit cumbersome to implement this ?

adityaghantasala committed:
Adding MasterSlaveConstraint::Pointer as a container in python

adityaghantasala committed:
formatting and local variable name changes

adityaghantasala committed:
removing commented lines of code

adityaghantasala committed:
Adding the type of returned vairable for the functions and formatting changes

adityaghantasala committed:
Adding small intial test to for constraints

adityaghantasala committed:
small code optimizations, clear comments and adding default arguments in constructors

adityaghantasala committed:
Adding copy constructor

adityaghantasala committed:
adding more details to the test

RiccardoRossi commented on kratos/python/add_model_part_to_python.cpp:
```diff
@@ -870,9 +870,9 @@ void AddModelPartToPython(pybind11::module& m)
-        .def("CeateNewMasterSlaveConstraint",CreateNewMasterSlaveConstraint1)
-        .def("CeateNewMasterSlaveConstraint",CreateNewMasterSlaveConstraint2)
-        .def("CeateNewMasterSlaveConstraint",CreateNewMasterSlaveConstraint3)
+        .def("CeateNewMasterSlaveConstraint",CreateNewMasterSlaveConstraint1, return_value_policy::reference_internal)
```
i think this should give back a pointer no?

adityaghantasala committed:
changing the function exposed to core

adityaghantasala committed:
removing the master_slave_relation.h from this branch

adityaghantasala committed:
Adding MasterSlaveConstraintsArray to python to facilitate access of the array of constraints from python

adityaghantasala committed:
Adding baseclass to python exposing, also adding print functionality.

adityaghantasala committed:
Adding PrintInfo function definition

adityaghantasala committed:
Adding assert statement to check if constraint is present in the container of MasterSlaveConstraints or not

RiccardoRossi commented on kratos/tests/test_model_part.py:
```diff
@@ -659,6 +659,20 @@ def test_model_part_has_solution_step_variable(self):
+        c1 = MasterSlaveConstraint(10)
+        c2 = model_part.CeateNewMasterSlaveConstraint("MasterSlaveConstraint", 1, n1, PRESSURE, n2, PRESSURE, 0.5, 0.0)
+
+        model_part.AddMasterSlaveConstraint(c1)
```
good start... but i am afaid the testing should be way more extensive, to include deleting and adding to submodelparts and subsubmodelparts...

RiccardoRossi commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,401 @@
+    ///@{
+    friend class Serializer;
+
+    virtual void save(Serializer &rSerializer) const override
```
this is still empty and should be implemented before merging

RiccardoRossi replied:
also we should soon test it. @philbucher we'll need your help in the saving loading

philbucher replied:
I will take a look

adityaghantasala committed:
removing includes not in this commit anymore.

adityaghantasala committed:
Adding definition of Clear function

RiccardoRossi committed:
some imporovement - in particular making the base class of constaint to contain nothing, and deriving one class from it

RiccardoRossi committed:
new constraint class which contains user provided data

RiccardoRossi committed:
registering constraints

RiccardoRossi committed:
some more testing of constraints - still incomplete and still to be debugged

RiccardoRossi reviewed (commented):
ok, i did some changes myself. The tests still need to be improved, and i put a todo since i also find something that needs to be debugged

RiccardoRossi commented:
ok, from my point of view this is getting ready for being merged. the todos now are to enrich a little the testing and to debug outstanding problems. Other than that the "Create" mechanism for MultiPointConstraints should be reviewed.

philbucher commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,293 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+ModelPart::MasterSlaveConstraintType::Pointer ModelPart::pGetMasterSlaveConstraint(ModelPart::IndexType MasterSlaveConstraintId)
+{
+        auto i = mMasterSlaveConstraints.find(MasterSlaveConstraintId);
+        KRATOS_DEBUG_ERROR_IF(i == mMasterSlaveConstraints.end()) << " constraint index not found: " << MasterSlaveConstraintId << ".";
```
Should this be a debug error only?
The effects of this being unnoticed are quite severe I'd say => consider doing this check always
(I haven't checked how it is done for elems/ conds)

adityaghantasala replied:
You are correct ... It should be done always. I change it.

RiccardoRossi replied:
let me disagree on this. It is like the [] operator on a std vector. the check is only done in debug...

philbucher commented on kratos/includes/user_provided_linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,319 @@
+    *                    : SlaveEqID = weight1 * MasterEqId1 + weight2 * MasterEqId2 + Constant1+Constant2
+    * This unique equation is used later on to modify the equation system.
+    */
+class UserProvidedLinearMasterSlaveConstraint :  public MasterSlaveConstraint
```
This seems like a strange name to me ...
Why not `LinearMasterSlaveConstraint`?

RiccardoRossi replied:
Ok for me to change it

philbucher commented on kratos/includes/user_provided_linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,319 @@
+
+
+
+  private:
```
this class also needs serialization

philbucher reviewed (commented):
Some comments

Big question:
From what I see currently the constraints have to be applied manually
What is the plan for the future?
Through processes I guess, or even the mdpa?

philbucher commented on kratos/tests/test_model_part.py:
```diff
@@ -659,6 +659,47 @@ def test_model_part_has_solution_step_variable(self):
+        sub2 = model_part.CreateSubModelPart("sub2")
+        subsub1 = sub1.CreateSubModelPart("subsub1")
         
+        ss1 = subsub1.CeateNewMasterSlaveConstraint("UserProvidedLinearMasterSlaveConstraint", 2, n1, PRESSURE, n2, PRESSURE, 0.5, 0.0)
```
If this works this is a serious typo!

philbucher commented on kratos/includes/user_provided_linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,319 @@
+
+namespace Kratos
+{
+/** \brief Constraint * @class MasterSlaveRelation
```
I think this is outdated

msandre replied:
+1

adityaghantasala replied:
you mean the description in the comment or the format ?

philbucher commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,345 @@
+    /**
+	* Clears the maps contents
+	*/
+    void Clear()
```
virtual?

philbucher commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,345 @@
+    MasterSlaveConstraint(const MasterSlaveConstraint& rOther)
+    {
+        this->SetId(rOther.Id());
+        // this->Flags = rOther.Flags;
```
commented?

philbucher commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,345 @@
+    {
+
+    }
+
```
Since this class is the baseclass I would say it also needs an assignment operator

msandre replied:
It might be the safer option to disable assignment for a virtual base class since it is not guaranteed that the derived object will be correctly assigned. It might not even be the same type. For the same reason I would consider removing the copy constructor.

philbucher commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,345 @@
+     * is called to initialize the constraint
+     * if the constraint needs to perform any operation before any calculation is done
+     */
+    virtual void Initialize()
```
don't you also need `Finalize`?

philbucher reviewed (commented):
more comments

philbuchersimscale committed:
Adding serialization for constraints

philbuchersimscale committed:
minor cleaning

philbucher commented on kratos/sources/model_part.cpp:
```diff
@@ -855,6 +855,293 @@ void ModelPart::RemoveElementsFromAllLevels(Flags identifier_flag)
+MasterSlaveConstraint::Pointer ModelPart::CreateNewMasterSlaveConstraint(const std::string& ConstraintName, 
+                                                                                    ModelPart::IndexType Id, 
+                                                                                    ModelPart::NodeType& rMasterNode,
+                                                                                    const ModelPart::DoubleVariableType& rMasterVariable,
```
Are the Master and the Slavevariable Historical variables?

If so, then I would add a debug check if these variables are in the modelpart

EDIT or probably it is better to check if a dof for the variable exists on the node

philbuchersimscale committed:
Testing serializeation of constraints in ModelPart

philbuchersimscale committed:
dummy c test linear constraint

philbucher commented:
I added the serialization related to the constraints

I also added a dummy c-test

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+//
+//
+
+#if !defined(CONSTRAINT_H)
```
You should use the name of the header file for this.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+#define CONSTRAINT_H
+// System includes
+
+// project inclu
```
small typo.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+#include "includes/process_info.h"
+namespace Kratos
+{
+/** \brief Constraint * @class MasterSlaveRelation
```
Clean up doxygen. Class name is wrong.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+{
+
+  public:
+    /// Pointer definition of DataValueContainer
```
Remove or fix comment.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+    ///@{
+
+   /*
+    * Empty Constructor
```
Use `///`

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+        // this->Flags = rOther.Flags;
+    }
+
+    // template <typename ...Params>
```
Better to remove commented blocks like this.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+
+    /**
+     * creates a new constraint pointer
+     * @param NewId the ID of the new element
```
element?

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+    /**
+     * creates a new constraint pointer
+     * @param NewId the ID of the new element
+     * @param ThisNodes the nodes of the new element
```
Doxygen params don't match function.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    /**
+     * this is called in the beginning of each solution step
```
You don't need to comment what is clearly expressed in the function name.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+
+    /**
+     * this determines the master equation IDs connected to this constraint
+     * @param rResult the elemental equation ID vector
```
elemental?

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+
+    /**
+     * this is called during the assembling process in order
+     * to calculate all elemental contributions to the global system
```
elemental?

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+                                      VectorType& rConstantVector,
+                                      ProcessInfo& rCurrentProcessInfo)
+    {
+      if (rTransformationMatrix.size1() != 0)
```
Why is the rConstantVector not resized with the matrix?

msandre commented on kratos/includes/model_part.h:
```diff
@@ -574,6 +604,110 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+    */
+    void RemoveMasterSlaveConstraintFromAllLevels(MasterSlaveConstraintType& ThisMasterSlaveConstraint);
+
+    /** Returns the MasterSlaveConstraint::Pointer  corresponding to it's identifier */
```
Another example where you could skip the comment.

msandre commented on kratos/includes/model_part.h:
```diff
@@ -574,6 +604,110 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+        return mMasterSlaveConstraints;
+    }
+
+    const MasterSlaveConstraintContainerType& MasterSlaveConstraints() const 
```
@pooyan-dadvand or @RiccardoRossi is it going to cause any problems that we don't use a mesh index like for elements and conditions?

RiccardoRossi replied:
constraints are not in mesh...so we simply cannot add it to a mesh.

if your case is about adding constraint to the mesh instead of to the modelpart directly, than you may be right...simply i sort of consider the meshes as legacy (my personal opinion on this though...)

pooyan-dadvand replied:
I see the @RiccardoRossi point of view but at this moment I would add them to the mesh like other entities in Modelpart to be consistent. Also consider that communicator works with meshes and having constraints outside would imply a parallel house keeping structure

msandre commented on kratos/includes/user_provided_linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,322 @@
+class UserProvidedLinearMasterSlaveConstraint :  public MasterSlaveConstraint
+{
+public:
+    /// Pointer definition of DataValueContainer
```
Fix or remove comment.

msandre commented on kratos/python/add_constraint_to_python.h:
```diff
@@ -0,0 +1,40 @@
+
+
+#if !defined(KRATOS_ADD_CONSTRANT_TO_PYTHON_H_INCLUDED )
+#define  KRATOS_ADD_CONSTRANT_TO_PYTHON_H_INCLUDED
```
typo in the def.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,344 @@
+     * @param rConstant the elemental right hand side
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    virtual void CalculateLocalSystem(MatrixType& rTransformationMatrix,
```
The transformation matrix is different from the local system right? Could we name this function differently?

adityaghantasala replied:
Thought the purpose of transformation matrix is different, it is also kind of local system. This is like  the provided Ke to the element into the constructor. 
I kept that name to be same as Element/Condition.

msandre commented on kratos/includes/user_provided_linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,322 @@
+     * to calculate all elemental contributions to the global system
+     * matrix and the right hand side
+     * @param rTransformationMatrix the elemental left hand side matrix
+     * @param rConstant the elemental right hand side
```
It's better if we don't comment overrides. The comment can be found in the base method and we avoid diverging / wrong comments.

msandre commented on kratos/tests/test_model_part.py:
```diff
@@ -659,6 +659,47 @@ def test_model_part_has_solution_step_variable(self):
+        self.assertFalse(ss1 in sub1.MasterSlaveConstraints)
+        self.assertTrue(ss1 in model_part.MasterSlaveConstraints)
+
+        subsub1.RemoveMasterSlaveConstraintFromAllLevels(ss1)
```
Why remove a second time from subsub1 and not check anything?

RiccardoRossi replied:
next line does an assertFalse which is true in 696

msandre commented on kratos/tests/test_model_part.py:
```diff
@@ -659,6 +659,47 @@ def test_model_part_has_solution_step_variable(self):
+        VariableUtils().AddDof(PRESSURE,model_part)
+
+        c1 = MasterSlaveConstraint(10)
+        c2 = model_part.CeateNewMasterSlaveConstraint("UserProvidedLinearMasterSlaveConstraint", 1, n1, PRESSURE, n2, PRESSURE, 0.5, 0.0)
```
c2 is not used anywhere.

adityaghantasala replied:
yes it is not. this is assigned as the function returns an object. I didn't want to leave the object unassigned.

msandre replied:
Ok, the tests could be improved by separating them into appropriately named test functions rather than mixing them in one function. Tests of functions intended to be used in c++ could be moved to the c++ test suite. Also, they could be more complete. For example, you create c2 with some values for the weight and constant. Why not test that CalculateLocalSystem returns the values expected?

philbucher replied:
I created a dummy c-test to test exactly that
=> https://github.com/KratosMultiphysics/Kratos/blob/8bc6511025f1fb9455693b5b23baed5a5241b314/kratos/tests/sources/test_linear_master_slave_constraint.cpp

adityaghantasala replied:
This means this test will be heavy. That is a modelpart has to be created, a few nodes, variables and dofs. Will that be ok for the KratosCoreFastSuite ?

philbucher replied:
this is done all the time, so its ok for the fastsuite

msandre replied:
Ideally, there would be one test per function. Each test should be minimal in the sense that it doesn't produce more work than is necessary to test that particular behavior. You can write a small test model factory to use in each of the unit tests.

msandre reviewed (changes requested):
The comments should be cleaned before merging. I tagged a few examples.

adityaghantasala commented:
@philbucher  Yes you are correct the constraints are to be formulated and implemented in a process. Not in mdpa.

adityaghantasala committed:
Merge branch 'feature-modelpart-interface-constraints' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

Changing from the KRATOS_DEBUG_IF to KRATOS_ERROR_IF

adityaghantasala committed:
Adding comments update and virtual to clear

adityaghantasala committed:
more updates to the doxygen comments and resizing const vector in the base class method

adityaghantasala committed:
bug fix

adityaghantasala committed:
Adding checks for master and slave variables on the master and slave nodes in create master slave constraint function

adityaghantasala committed:
typo fix

adityaghantasala commented:
@philbucher  I the serialization you implemented in the user_provided_linear_master_slave_constraint.h might be problematic. This is because, 
- The mSlaveDofsVector is vector of shared pointers. 
- I am not sure if we can save these dofs here or not as they can be reconstructed after a restart.

RiccardoRossi commented:
@adityaghantasala i would use raw pointers to the dofs instead of shared ptrs.

in any case this should not be a problem for the serialization... (pending that we test it of course)

adityaghantasala commented:
@RiccardoRossi  when storing the raw pointers : if the pointer is cleared twice there will be a problem I think. Because the dof is deleted by the node and also the destructor of the linear master slave constraint. 
One question :  when a pointer is passed to the serializer the object behind the pointer is serialized right ?

adityaghantasala committed:
formatting and removing the spaces in test_model_part.py
adding implementation in c++  test

adityaghantasala committed:
Fixing remove_constraints_from_all_levels

adityaghantasala committed:
adding back the check if the constraints are removed from all the levels

adityaghantasala commented:
@RiccardoRossi  The latest push solves the TODO: you put in the test. There was mismatch of function calls in model_part.cpp. I fixed it and it works now.

adityaghantasala committed:
typo correction in adding creat function to python

adityaghantasala committed:
correction of funciton call for typo

adityaghantasala committed:
addin missing over load for create function in user defined linear master slave constraint

adityaghantasala committed:
Adding GetDofList function to base and derived master slave constraint class

adityaghantasala committed:
removing the doxigen comments from derived master slave constraint class

adityaghantasala committed:
bug fixes

adityaghantasala committed:
small fix in variable name usage

adityaghantasala committed:
adding missing override in create function definion

RiccardoRossi reviewed (commented):
@KratosMultiphysics/technical-committee i think this is ready for merging. There are still minor stylistic comments (like removing comments and commented code, @msandre please just take a look if this is ok or if your comments are still valid) 

...however overall i would merge this. I am waiting for someone else in @KratosMultiphysics/technical-committee to take a look into it before proceeding

adityaghantasala committed:
Changing name from UserProvidedLinearMasterSlaveConstraint to LinearMasterSlaveConstraint.
Necessary changes propagated

adityaghantasala committed:
removing copy constructor from base class of master slave constraint

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
small update to test modelpart to use the un used variables.

adityaghantasala committed:
Adding add_constraints_to_python.cpp to sources

adityaghantasala committed:
small change in @brief of the masterslaveconstraint class

Commit:
Cleaned doxygen.

Commit:
Removed unused includes / typedefs.

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Merge branch 'feature-modelpart-interface-constraints' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Avoiding name conflicts : redefining helper function names

adityaghantasala commented:
ping !!

msandre commented:
This should be ready for approval. ping @KratosMultiphysics/technical-committee.

RiccardoRossi commented:
i am adding this to the todo for tomorrow's meeting of @KratosMultiphysics/technical-committee

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Adding missing AddMasterSlaveConstraints to python.
Adding missing AddMasterSlaveConstraints (using iterators)

pooyan-dadvand commented on kratos/includes/model_part.h:
```diff
@@ -584,6 +614,157 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+        return mMasterSlaveConstraints;
+    }
+
+    void SetMasterSlaveConstraints(const MasterSlaveConstraintContainerType& rOtherMasterSlaveConstraints)
```
Do we need this method?

adityaghantasala replied:
I needed this to define MasterSlaveConstraints as a property in the python interface. (a Set and Get function is required isn't it ?)

pooyan-dadvand commented on kratos/includes/model_part.h:
```diff
@@ -584,6 +614,157 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+
+    void SetMasterSlaveConstraints(const MasterSlaveConstraintContainerType& rOtherMasterSlaveConstraints)
+    {
+        mMasterSlaveConstraints = rOtherMasterSlaveConstraints;
```
shouldn't we add the constrains to the parent modelpart?

pooyan-dadvand commented on kratos/includes/model_part.h:
```diff
@@ -1347,6 +1528,8 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
     SubModelPartsContainerType mSubModelParts;
 
+    MasterSlaveConstraintContainerType mMasterSlaveConstraints;
+
```
You may move this to the mesh

adityaghantasala replied:
This means all the interface functions should also be added to the mesh class ?

pooyan-dadvand commented on kratos/includes/model_part.h:
```diff
@@ -584,6 +614,157 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
+    MasterSlaveConstraintType::Pointer pGetMasterSlaveConstraint(IndexType ConstraintId);
+
+    /** Returns a reference MasterSlaveConstraint corresponding to it's identifier */
+    MasterSlaveConstraintType& GetMasterSlaveConstraint(IndexType MasterSlaveConstraintId);
```
get methods should have a const overload

adityaghantasala replied:
done

pooyan-dadvand commented on kratos/python/add_constraint_to_python.cpp:
```diff
@@ -0,0 +1,62 @@
+{
+namespace Python
+{
+using namespace pybind11;
```
I would suggest to make the interface without using namespace

adityaghantasala replied:
Done !

pooyan-dadvand commented on kratos/tests/sources/test_linear_master_slave_constraint.cpp:
```diff
@@ -0,0 +1,65 @@
+        KRATOS_CHECK_EQUAL(transformation_matrix.size2(), 1);
+        KRATOS_CHECK_EQUAL(constant_vector.size(), 1);
+
+        KRATOS_CHECK_EQUAL(transformation_matrix(0,0), 1.0); // TODO: Check -> comparison between two doubles ??
```
you may use `KRATOS_CHECK_DOUBLE_EQUAL`

pooyan-dadvand reviewed (commented):
There are some minor comments. The most important ones are:
- The set method for modelpart which is not sub model part compatible
- Including the constraints in the mesh?

adityaghantasala committed:
Changing to KRATOS_CHECK_DOUBLE_EQUAL

adityaghantasala committed:
Adding const overload of the GetMasterSalaveConstraint

adityaghantasala committed:
moving the MasterSlaveConstraint container to the Mesh class.
Necessary changes and propagation of changes to the necessary files.
redefines Mesh class template

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

pooyan-dadvand commented:
@adityaghantasala do you need this to be opened or can be deleted?

adityaghantasala commented:
@pooyan-dadvand  this is not yet merged. I am making the changes we discussed yesterday once done, I will merge it and delete it.

adityaghantasala committed:
Removie MasterSlaveConstraintType as a template perameter in Mesh class

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
Exposing the MasterSlaveConstraints only a read only data to the python

adityaghantasala committed:
removing SetMasterSlaveConstraints in from the modelpart and implementation

msandre commented on kratos/includes/linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,299 @@
+    ///@name Life Cycle
+    ///@{
+
+    LinearMasterSlaveConstraint(IndexType Id = 0)
```
Add `explicit`

msandre commented on kratos/includes/linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    /// Copy Constructor
+    LinearMasterSlaveConstraint(const LinearMasterSlaveConstraint& rOther)
```
Add an assignment operator if you have copy constructor.

msandre commented on kratos/includes/linear_master_slave_constraint.h:
```diff
@@ -0,0 +1,299 @@
+                                const VectorType& rConstantVector)
+        : MasterSlaveConstraint(Id)
+    {
+        mSlaveDofsVector = rSlaveDofsVector;
```
Better to initialize these in the initialization list. By assigning in the body you may prevent move ops from happening automatically.

msandre commented on kratos/includes/master_slave_constraint.h:
```diff
@@ -0,0 +1,377 @@
+
+
+    /// Empty Constructor
+    MasterSlaveConstraint(IndexType Id = 0) : IndexedObject(Id), Flags()
```
Add `explicit`

msandre reviewed (changes requested):
There are a few small things to include in the final cleaning before merge.

adityaghantasala committed:
explicit keyword for constructor,  assignment operator for linear master slave constraint.

adityaghantasala committed:
initializing members in initializer list

adityaghantasala committed:
not using namespace pybind11

adityaghantasala committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into feature-modelpart-interface-constraints

adityaghantasala committed:
removing explicit from overloaded constructors

msandre reviewed (approved):
Approving from my side. After last week's discussion, I think there are no more changes requested for this PR. Thanks @adityaghantasala!

adityaghantasala commented:
Merging this branch as discussed last week in person with @RiccardoRossi and @pooyan-dadvand

philbucher commented:
:+1: 

Do you still need this branch?

pooyan-dadvand commented:
@adityaghantasala do you need the branch to be alive?

adityaghantasala commented:
no ... deleting this ... !
